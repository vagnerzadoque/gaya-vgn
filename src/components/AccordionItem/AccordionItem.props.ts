import React from 'react'
import { BrandTypes } from '../../brandTypes/brandTypes'

export interface AccordionItemProps {

  brand?: BrandTypes
  /**
   * Specify an optional className to be added to your AccordionItem
   */
  className?: string

  /**
   * Optional ID for testing
   */
  testID?: string;

  /**
   * Set text header of AccordionItem
   */
  title: string

  /**
   * Optional text usign as "subtitle"
   */
  legend?: string

  /**
   * Parameter to define the style that will be applied
   * @default info
  */
  color?: 'regular' | 'primary'

  /**
   * Define state of item (open/close)
   * @default false
  */
  isActive?: boolean

  /**
   * If "true" disable event click and change color style
   * @default false
  */
  isDisabled?: boolean

  /**
   * Function to toggle state between open and close
  */
  onClick: () => void

  /**
   * Content will be render when AccordionItem is open
  */
  children: React.ReactNode
}
