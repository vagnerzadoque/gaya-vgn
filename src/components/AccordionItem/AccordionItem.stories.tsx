import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { fn } from '@storybook/test'
import { useState } from 'react'
import StoryContainer from '../../helpers/StoryContainer'
import { AccordionItem, AccordionItemProps } from '.'

const componentStatus = `
> AccordionItem representing an individual item that can be expanded or collapsed in response to user interaction.

## Properties
| Property         | Values                |    Status           |
|---               |                    ---|                  ---|
| **variant**      | standard              | ✅ Available        |
| **title**        | string                | ✅ Available        |
| **legend**       | string                | ✅ Available        |
| **color**        | regular, primary'     | ✅ Available        |
| **isActive**     | true/false            | ✅ Available        |
| **isDisabled**   | true/false            | ✅ Available        |
| **onClick**      | function              | ✅ Available        |
| **brand**        | avon, avon_v2, natura, natura_v2, theBodyShop, <br /> consultoriaDeBeleza, casaEestilo              | ✅ Available        |


## Technical Usages Examples
`

const meta: Meta<AccordionItemProps> = {
  title: 'Components/AccordionItem',
  component: AccordionItem,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<AccordionItemProps>;

export const Playground: Story = {
  args: {
    title: "Subtitle 1",
    children: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Modi deserunt, ratione alias recusandae error fuga vero obcaecati impedit in dicta esse ad debitis iure voluptatem ipsam atque! Quam, doloribus alias!"
  },
  render: (args) => {
    const [isActive, setIsActive] = useState(false)
    const onClick = () => setIsActive(!isActive)

    return (
      <StoryContainer>
        <AccordionItem {...args} isActive={isActive} onClick={onClick} />
      </StoryContainer>
    )
  }
};

export const Primary: Story = {
  args: {
    title: "Subtitle 1",
    color: "primary",
    children: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Modi deserunt, ratione alias recusandae error fuga vero obcaecati impedit in dicta esse ad debitis iure voluptatem ipsam atque! Quam, doloribus alias!"
  },
  render: (args) => {
    const [isActive, setIsActive] = useState(true)
    const onClick = () => setIsActive(!isActive)

    return (
      <StoryContainer>
        <AccordionItem {...args} isActive={isActive} onClick={onClick} />
      </StoryContainer>
    )
  }
};

export const Brand: Story = {
  args: {
    brand: "avon_v2",
    title: "Subtitle 1",
    color: "primary",
    children: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Modi deserunt, ratione alias recusandae error fuga vero obcaecati impedit in dicta esse ad debitis iure voluptatem ipsam atque! Quam, doloribus alias!"
  },
  render: (args) => {
    const [isActive, setIsActive] = useState(true)
    const onClick = () => setIsActive(!isActive)

    return (
      <StoryContainer>
        <AccordionItem {...args} isActive={isActive} onClick={onClick} />
      </StoryContainer>
    )
  }
};

export const Disabled: Story = {
  args: {
    title: "Subtitle 1",
    isDisabled: true,
    children: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Modi deserunt, ratione alias recusandae error fuga vero obcaecati impedit in dicta esse ad debitis iure voluptatem ipsam atque! Quam, doloribus alias!"
  },
  render: (args) => (
    <StoryContainer>
      <AccordionItem {...args} onClick={fn()} />
    </StoryContainer>
  )
};

export const Legend: Story = {
  args: {
    title: "Subtitle 1",
    legend: "Legend",
    children: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Modi deserunt, ratione alias recusandae error fuga vero obcaecati impedit in dicta esse ad debitis iure voluptatem ipsam atque! Quam, doloribus alias!"
  },
  render: (args) => {
    const [isActive, setIsActive] = useState(true)
    const onClick = () => setIsActive(!isActive)

    return (
      <StoryContainer>
        <AccordionItem {...args} isActive={isActive} onClick={onClick} />
      </StoryContainer>
    )
  }
};
