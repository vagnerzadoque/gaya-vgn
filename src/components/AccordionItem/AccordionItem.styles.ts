/* eslint-disable no-nested-ternary */
/* eslint-disable complexity */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { buildTheme } from '../../ThemeProvider'
import { BrandTypes } from '../../brandTypes/brandTypes'

type StyleProps = {
  color: 'regular' | 'primary',
  isActive: boolean,
  isDisabled: boolean,
  brand: BrandTypes | undefined
}

const overlay = (theme: Theme) => ({
  content: '" "',
  position: 'absolute',
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
  backgroundColor: theme.color.highlight,
  opacity: theme.opacity.lower
})

const applyOpenStyle = (theme: Theme) => ({ isActive }: StyleProps) => {
  if (isActive) {
    return overlay(theme)
  }

  return {}
}

const applyClickStyle = (theme: Theme) => {
  const baseOverlay = overlay(theme)
  baseOverlay.opacity = theme.opacity.veryLow

  return baseOverlay
}

const setBackgroundColor = (theme: Theme) => ({ isDisabled, color, brand }: StyleProps) => {
  const themeBrandSelect = buildTheme(brand, 'light')

  if (isDisabled) {
    return theme.color.lowEmphasis
  }

  return color === 'regular' ? theme.color.surface : (brand ? themeBrandSelect.color.primary : theme.color.primary)
}

const setLegendColor = (theme: Theme) => ({ isDisabled, color }: StyleProps) => {
  if (isDisabled) {
    return theme.color.mediumEmphasis
  }

  return color === 'regular' ? theme.color.mediumEmphasis : theme.color.highEmphasis
}
const setHeaderColor = (theme: Theme) => ({ isDisabled, color, brand }: StyleProps) => {
  const themeBrandSelect = buildTheme(brand, 'light')
  if (isDisabled) {
    return theme.color.mediumEmphasis
  }

  return color === 'regular' ? theme.color.mediumEmphasis : (brand ? themeBrandSelect.color.onPrimary : theme.color.onPrimary)
}

const styles = createUseStyles((theme: Theme) => ({
  wrapper: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
    background: theme.color.surface,
    fontFamily: theme.typography.fontFamily.primary,
    color: ({ isDisabled }: StyleProps) => (
      isDisabled ? theme.color.mediumEmphasis : theme.color.highEmphasis),
    pointerEvents: ({ isDisabled }: StyleProps) => (isDisabled ? 'none' : 'auto')
  },
  header: (props: StyleProps) => ({
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
    padding: `${theme.spacing.small}px ${theme.spacing.standard}px`,
    position: 'relative',
    color: setHeaderColor(theme)(props),
    background: setBackgroundColor(theme)(props),
    '& > div': {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between',
      fontWeight: theme.typography.fontWeight.medium
    },
    '& > span': {
      fontSize: theme.typography.fontSize.level1,
      marginTop: theme.spacing.small
    },
    '&:before': {
      ...applyOpenStyle(theme)(props)
    },
    '&:active:after': {
      ...applyClickStyle(theme)
    },
    '&:hover': {
      cursor: 'pointer'
    },
    '&:hover:before': {
      ...overlay(theme)
    }
  }),
  legend: {
    color: setLegendColor(theme)
  },
  content: {
    minHeight: theme.size.semiX,
    padding: `${theme.spacing.small}px ${theme.spacing.standard}px`,
    fontWeight: theme.typography.fontWeight.regular
  }
}))

export default styles
