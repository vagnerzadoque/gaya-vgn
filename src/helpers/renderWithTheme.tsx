import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
