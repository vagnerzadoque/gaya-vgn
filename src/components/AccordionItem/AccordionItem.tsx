import React from 'react'
import { Color } from '@naturacosmeticos/natds-themes'
import styles from './AccordionItem.styles'
import { AccordionItemProps } from './AccordionItem.props'
import Icon from '../Icon'

const AccordionItem = React.forwardRef<HTMLDivElement, AccordionItemProps>(
  (props, ref) => {
    const {
      className = '',
      title,
      brand,
      legend: legendText,
      color = 'regular',
      isActive = false,
      isDisabled = false,
      onClick,
      children,
      ...rest
    } = props
    const {
      wrapper, header, content, legend
    } = styles({
      color, isActive, isDisabled, brand
    })

    const setIconColor = (): keyof Color => {
      if (isDisabled) {
        return 'mediumEmphasis'
      }

      return color === 'regular' ? 'primary' : 'onPrimary'
    }

    return (
      <div
        className={`${className} ${wrapper} AccordionItem-gaya`}
        ref={ref}
        {...rest}
      >
        <div className={header} onClick={onClick}>
          <div>
            {title}
            <Icon
              brand={brand}
              name={
            isActive
              ? 'outlined-navigation-arrowtop'
              : 'outlined-navigation-arrowbottom'
}
              size="standard"
              color={setIconColor()}
            />
          </div>

          {legendText && <span className={legend}>{legendText}</span>}
        </div>

        {isActive && (
        <div className={content}>
          {children}
        </div>
        )}
      </div>
    )
  }
)

export default AccordionItem
