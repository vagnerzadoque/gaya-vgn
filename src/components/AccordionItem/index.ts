import AccordionItem from './AccordionItem'

export { AccordionItem }
export { AccordionItemProps } from './AccordionItem.props'
export default AccordionItem
