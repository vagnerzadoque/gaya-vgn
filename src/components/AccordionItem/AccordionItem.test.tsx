import userEvent from '@testing-library/user-event'
import React from 'react'
import { fireEvent } from '@testing-library/react'
import AccordionItem from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

const onClickSpy = jest.fn()

describe('AccordionItem component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render component with regular style as default', () => {
    const { styles, component } = renderWithTheme(
      <AccordionItem title="Lorem" onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should calls onClick when header is clicked', () => {
    const { component } = renderWithTheme(
      <AccordionItem title="Header title" onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    // Using fireEvent instead of userEvent for more direct event triggering
    fireEvent.click(component.getByText(/header title/i))

    expect(onClickSpy).toHaveBeenCalled()
  })

  it('should disabled clicks when is disabled (regular)', () => {
    const { styles } = renderWithTheme(
      <AccordionItem title="Header title" isDisabled onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    expect(styles.toString()).toContain('pointer-events: none')
  })

  it('should disabled clicks when is disabled (primary)', () => {
    const { styles } = renderWithTheme(
      <AccordionItem color="primary" title="Header title" isDisabled onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    expect(styles.toString()).toContain('pointer-events: none')
  })

  it('should toggle item when is active is true', () => {
    const { component } = renderWithTheme(
      <AccordionItem title="Header title" isActive onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    expect(component.getByTestId('icon-outlined-navigation-arrowtop')).toBeInTheDocument()
  })

  it('should render legend when is passed', () => {
    const { component } = renderWithTheme(
      <AccordionItem title="Header title" legend="Legend text" isActive onClick={onClickSpy}>Accordion Content</AccordionItem>
    )

    expect(component.getByText(/legend text/i)).toBeInTheDocument()
  })
})
