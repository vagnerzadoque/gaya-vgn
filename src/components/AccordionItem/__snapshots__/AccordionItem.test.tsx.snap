// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem component should render component with regular style as default 1`] = `
[
  ".wrapper-0-2-1 {
  flex: 1;
  display: flex;
  background: #ffffff;
  font-family: Roboto;
  flex-direction: column;
}
.header-0-2-2 {}
.legend-0-2-3 {}
.content-0-2-4 {
  padding: 16px 24px;
  min-height: 40px;
  font-weight: 400;
}
.wrapper-d0-0-2-5 {
  color: #363636;
  pointer-events: auto;
}
.header-d1-0-2-6 {
  flex: 1;
  color: #737373;
  display: flex;
  padding: 16px 24px;
  position: relative;
  background: #ffffff;
  flex-direction: column;
}
.header-d1-0-2-6 > div {
  display: flex;
  align-items: center;
  font-weight: 500;
  justify-content: space-between;
}
.header-d1-0-2-6 > span {
  font-size: 13px;
  margin-top: 16px;
}
.header-d1-0-2-6:before {}
.header-d1-0-2-6:active:after {
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  content: " ";
  opacity: 0.08;
  position: absolute;
  background-color: #111111;
}
.header-d1-0-2-6:hover {
  cursor: pointer;
}
.header-d1-0-2-6:hover:before {
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  content: " ";
  opacity: 0.04;
  position: absolute;
  background-color: #111111;
}
.legend-d2-0-2-7 {
  color: #737373;
}
.icon-0-2-8 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-9 {
  color: #f4ab34;
  font-size: 24px;
}",
  <div>
    <div
      class=" wrapper-0-2-1 wrapper-d0-0-2-5 AccordionItem-gaya"
    >
      <div
        class="header-0-2-2 header-d1-0-2-6"
      >
        <div>
          Lorem
          <i
            aria-hidden="true"
            class=" icon-0-2-8 icon-d0-0-2-9 natds-icons natds-icons-outlined-navigation-arrowbottom Icon-gaya"
            data-testid="icon-outlined-navigation-arrowbottom"
            role="img"
          />
        </div>
      </div>
    </div>
  </div>,
]
`;
